feat(quantic): logic added to properly handle the initialization errors in the Quantic components

mmitiche opened the pull request:
[SFINT-4946](https://coveord.atlassian.net/browse/SFINT-4946)

In this PR:

- Created the Quantic Component Error component.
- Updated the `initializeWithHeadless` method of the QuanticHeadlessLoader to avoid breaking the whole initialization loop when ever an error occurs while initializing one of the components.
- the `initializeWithHeadless` method now uses the [Lightning Alert 
](https://developer.salesforce.com/docs/component-library/bundle/lightning-alert/documentation) to display a modal containing an error message whenever one of the components fails to initialize.
- The component that fails to initialize now displays an error message using the new  Quantic Component Error, I updated all the Quantic components that use Headless controllers with this logic.
- The Quantic document suggestion and the Quantic case classification were already rendering error messages when invalid properties were passed to these component, so just adapted their logic to work with the new Quantic Error Component.
- Some components like the Quantic Pager, the Quantic Result Per Page and the Quantic Facet were already having some e2e tests to cover the case where invalid properties were passed, I adapted these tests with the new Quantic Error Component.




## Demo:

https://github.com/coveo/ui-kit/assets/86681870/560a8919-2a80-4eff-be3c-b0bbc92cdb12

## The new structure of the HTML templates of Quantic Components:
```
<template>
  <template if:true={hasInitializationError}>
    <c-quantic-component-error component-name={template.host.localName}>
    </c-quantic-component-error>
  </template>
  <template if:false={hasInitializationError}>
      <!-- USUAL TEMPLATE -->
  </template>
</template>

```



[SFINT-4946]: https://coveord.atlassian.net/browse/SFINT-4946?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

mmitiche commented on packages/quantic/force-app/main/default/lwc/quanticCaseClassification/quanticCaseClassification.js:
```diff
@@ -26,6 +35,9 @@ export default class QuanticCaseClassification extends LightningElement {
     loading,
+    salesforceFieldNotFound,
+    invalidMaxNumberOfSuggestions,
+    propertyisRequired,
```
```suggestion
    propertyIsRequired,
```

mmitiche commented on packages/quantic/force-app/main/default/lwc/quanticCaseClassification/quanticCaseClassification.js:
```diff
@@ -187,17 +205,26 @@ export default class QuanticCaseClassification extends LightningElement {
-      this.renderingError = 'coveoFieldName is required, please set its value.';
+      this.hasInitializationError = true;
+      this.initializationErrorMessage = `${I18nUtils.format(
+        this.labels.propertyisRequired,
```
```suggestion
        this.labels.propertyIsRequired,
```

nathanlb commented:
It's been a little while since we had that group discussion but was it not a good idea in the end to create a base component class the others could extend? Even if it just has hasComponentError and setComponentError it could same a lot of repetition.
Btw this is a cool addition to the lib! 🔥

mmitiche commented:
> It's been a little while since we had that group discussion but was it not a good idea in the end to create a base component class the others could extend? Even if it just has hasComponentError and setComponentError it could same a lot of repetition. Btw this is a cool addition to the lib! 🔥

The main reason I didn't go with the solution of having a base component class that the other Quantic components can extend is mainly because now the LWC component can extend from different classes and not only from `LightningElement`, for example we have some components that extend from `LightningModal` and some others extend from `NavigationMixin` and in the future we can have more cases like this, so it would be hard to maintain the solution of having one single base component that all the other quantic components will extend.

nathanlb commented:
> > It's been a little while since we had that group discussion but was it not a good idea in the end to create a base component class the others could extend? Even if it just has hasComponentError and setComponentError it could same a lot of repetition. Btw this is a cool addition to the lib! 🔥
> 
> The main reason I didn't go with the solution of having a base component class that the other Quantic components can extend is mainly because now the LWC component can extend from different classes and not only from `LightningElement`, for example we have some components that extend from `LightningModal` and some others extend from `NavigationMixin` and in the future we can have more cases like this, so it would be hard to maintain the solution of having one single base component that all the other quantic components will extend.

Ah yes thanks that's the bit of information i'd forgotten. Thanks!

SimonMilord commented on packages/quantic/force-app/main/default/labels/CustomLabels.labels-meta.xml:
```diff
@@ -1106,4 +1106,53 @@
+        <value>{{0}} component error</value>
+        <language>en_US</language>
+        <protected>false</protected>
+        <shortDescription>c-quantic-pager component error</shortDescription>
```
if the error applies to any component, why the c-quantic-pager in shortDescription?

mmitiche replied:
ah that's not the value that will be displayed, the value is in the tag `value`.
in the tag `shortDescription` I'm just adding an example of what the label should look like.

SimonMilord replied:
ahh ok ok

SimonMilord commented on packages/quantic/force-app/main/default/lwc/quanticTab/quanticTab.js:
```diff
@@ -98,7 +100,7 @@ export default class QuanticTab extends LightningElement {
   updateState() {
     this._isActive = this.tab?.state?.isActive;
-    this.shouldDisplay = this.searchStatus?.state?.firstSearchExecuted;
+    this.shouldDisplay = this.searchStatus?.state?.firstSearchExecuted && !this.hasInitializationError;
```
Question: wouldn't updateState() not be called in case of initializationError? Why check for init error at this point?

mmitiche replied:
in the case of an initialization error updateState is not called no, good point! I should keep them separate.

mmitiche replied:
Just rechecked, the logic here is the following:
We should display the tab only when the first search has been executed AND there are no error during initialization.

I'm including this condition about the error initialization here to avoid having to add an additional if:false template in the HTML.
 instead of doing: 
```
<template if:true={hasInitializationError}>Error</template>
<template if:false={hasInitializationError}>
    <template if:true={shouldDisplay}>Tab component</template>
</template>
```

I'm updating the shouldDisplay condition to be `this.searchStatus?.state?.firstSearchExecuted && !this.hasInitializationError;`
 in order to have:

```
<template if:true={hasInitializationError}>Error</template>
<template if:true={shouldDisplay}>Tab component</template>
```
Both are equivalent, the second simpler imo.

SimonMilord replied:
ah ok I see I get why the html file had no <template if:false={hasInitializationError}> then :-D

SimonMilord commented on packages/quantic/force-app/main/default/lwc/quanticTab/quanticTab.html:
```diff
@@ -1,4 +1,8 @@
 <template>
+  <template if:true={hasInitializationError}>
```
following my question in the js file, should we not have an `if:false` with the `if:true={shouldDisplay}` inside it?

SimonMilord commented on packages/quantic/force-app/main/default/lwc/quanticNumericFacet/quanticNumericFacet.js:
```diff
@@ -608,4 +608,11 @@ export default class QuanticNumericFacet extends LightningElement {
+  /**
+   * Sets the component in the initialization error state.
+   */
+  setInitializationError() {
```
I think the declaration for `hasInitializationError ` boolean is missing in this file?

SimonMilord commented on packages/quantic/cypress/page-objects/actions/action-perform-search.ts:
```diff
@@ -1,5 +1,5 @@
 export const performSearch = () =>
-  cy.get('c-action-perform-search button').click();
+  cy.get('c-action-perform-search button').click({force: true});
```
why are we setting `force: true` here?

mmitiche replied:
Cause when there is a component error, a modal dialog like you can see is displayed.
And in Cypress you can't click on something while a modal is open unless you set force to true.

SimonMilord commented on packages/quantic/cypress/e2e/default-1/results-per-page/results-per-page.cypress.ts:
```diff
@@ -125,6 +125,16 @@ describe('quantic-result-per-page', () => {
 
       describe('with invalid options', () => {
+        beforeEach(() => {
+          // This error occasionally occurs with the Salesforce Lightning Modal component, although we don't know exactly why it occurs, we do know that it only occurs in this Cypress test environment and never in the production environment.
```
technically also occurs in the pager tests :P

mmitiche replied:
what do you mean?

SimonMilord commented on packages/quantic/force-app/main/default/lwc/quanticComponentError/quanticComponentError.js:
```diff
@@ -0,0 +1,42 @@
+   * The error message to display.
+   * @api
+   * @type {string}
+   * @defaultValue `'Look at the developer console for more information.</'`
```
```suggestion
   * @defaultValue `'Look at the developer console for more information.'`
```

SimonMilord reviewed (approved):
Following my comments and the meeting we had about the PR, it looks good to me. Just remember to localize the labels and adding the suggestion for template Etienne was talking about would be cool. Good job though this is a cool addition to quantic indeed 🔥

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticHeadlessLoader/quanticHeadlessLoader.js:
```diff
@@ -305,7 +307,22 @@ async function initializeWithHeadless(element, engineId, initialize) {
+    element?.setInitializationError?.();
+    if (!componentErrorDisplayed) {
+      componentErrorDisplayed = true;
+      await LightningAlert.open({
+        message: `Fatal error: unable to initialize the ${element?.template?.host?.localName} component.
+        Look at the developer console for more information.`,
+        theme: 'error',
+        label: 'Component initialization error',
+      });
```
String revision and labelize these please :)

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticHeadlessLoader/quanticHeadlessLoader.js:
```diff
@@ -305,7 +307,22 @@ async function initializeWithHeadless(element, engineId, initialize) {
     initialize(await getHeadlessEnginePromise(engineId));
     setComponentInitialized(element, engineId);
```
You said you wanted to remove this since it's in the finally.

mrrajamanickam-coveo commented on packages/quantic/cypress/e2e/facets-1/facet/facet.cypress.ts:
```diff
@@ -787,7 +787,14 @@ describe('Facet Test Suite', () => {
-        before(() => {
+        beforeEach(() => {
+          // This error occasionally occurs with the Salesforce Lightning Modal component, although we don't know exactly why it occurs, we do know that it only occurs in this Cypress test environment and never in the production environment.
+          cy.on('uncaught:exception', (err) => {
```
Could this be refactored into a common place? If only certain tests need to have this then I guess this is better than having the common handler check for test names etc?

> To conditionally turn off uncaught exception handling for a certain error
A great place to put this configuration is in the [supportFile](https://docs.cypress.io/guides/core-concepts/writing-and-organizing-tests#Support-file), since it is loaded before any test files are evaluated.

mrrajamanickam-coveo replied:
> we do know that it only occurs in this Cypress test environment and never in the production environment. 

Is this based on production logs/monitoring or just that it hasn't been reported in prod?

mrrajamanickam-coveo commented on packages/quantic/cypress/e2e/facets-1/facet/facet.cypress.ts:
```diff
@@ -787,7 +787,14 @@ describe('Facet Test Suite', () => {
+            expect(err.message).to.include(
+              "Cannot read properties of null (reading 'appendChild')"
+            );
+            return false;
```
Adding a log (`cy.log()` e.g.) when this happens might be good to have for debugging etc.

mmitiche commented on packages/quantic/force-app/main/default/lwc/quanticHeadlessLoader/quanticHeadlessLoader.js:
```diff
@@ -305,7 +307,22 @@ async function initializeWithHeadless(element, engineId, initialize) {
+      error
+    );
+    element?.setInitializationError?.();
+    if (!componentErrorDisplayed) {
+      componentErrorDisplayed = true;
```
@erocheleau I'm using this variable to make sure I'm displaying the Error dialog only once even if you have errors in multiple components.

For example if you have invalid properties in both the Quantic Pager and the Quantic Result Per Page. You will get only one Error Modal for whichever component threw first, let's say it's the Quantic Pager, the message in the modal will be about the Quantic Pager. When you close the modal, there is no other modal that appears. However in the page the Error component will be displayed in both the Quantic Pager and the Quantic Result Per Page.

After fixing the invalid properties of the Quantic Pager and reloading the page, the error modal will appear aagain but this time containing an error message about the Quantic Result Per Page component.

erocheleau replied:
Makes total sense, good idea

mmitiche commented:
> As said during the meeting, explore if it would be possible instead to use the rendering method and to import the error template instead of adding an `if` to all templates.
> 
> Perhaps try with the BaseComponent too if it's possible to set an extend + super.defaultTemplate = ...
> 
> Also please post what happens if there are multiple errors are the same time? :) See if we get 10 modals lol

Added the localized labels and updated the directives used to  achieve conditional rendering in LWC.

mmitiche commented on packages/quantic/force-app/main/default/labels/CustomLabels.labels-meta.xml:
Hey @cmartel-coveo !
I would appreciate your input regarding the following labels.
These labels are displayed to the user when some invalid properties are passed to some Quantic components. Their main purpose is to inform the user about the error currently occurring in their Quantic search page.

Let me know if you need more details!
Thank you!

mmitiche commented on packages/quantic/force-app/main/translations/fr.translation-meta.xml:
@cmartel-coveo here you can find the french translations.
Thank you!

cmartel-coveo replied:
Hello @mmitiche !
Yep, got a few questions for you, I'll reach out on Slack, it'll be easier "in person".

cmartel-coveo commented on packages/quantic/force-app/main/default/labels/CustomLabels.labels-meta.xml:
```diff
@@ -1106,4 +1106,67 @@
+    </labels>
+    <labels>
+      <fullName>quantic_LookAtDeveloperConsole</fullName>
+      <value>Look at the developer console for more information.</value>
```
@MehdiMitiche Guessing that the developer console is on another screen, I would only change the verb at the beginning for: 
"Go to the developer console for more information."

cmartel-coveo commented on packages/quantic/force-app/main/default/labels/CustomLabels.labels-meta.xml:
```diff
@@ -1106,4 +1106,67 @@
+    </labels>
+        <labels>
+      <fullName>quantic_InvalidMaxNumberOfDocumentSuggestions</fullName>
+      <value>"{{0}}" is an invalid maximum number of document suggestions. An integer greater than 0 was expected.</value>
```
@MehdiMitiche Maybe I'm simplifying too much, but we could go straight to the solution, as 0 is obviously not valid as a maximum?

"The maximum number of document suggestions must be greater than 0."

mmitiche replied:
Totally agree!

cmartel-coveo commented on packages/quantic/force-app/main/default/labels/CustomLabels.labels-meta.xml:
```diff
@@ -1106,4 +1106,67 @@
+    </labels>
+    <labels>
+      <fullName>quantic_InvalidNumberOfAutoOpenedDocuments</fullName>
+      <value>"{{0}}" is an invalid number of automatically opened document suggestions. A positive integer was expected.</value>
```
@mmitiche Same idea here:

The number of automatically opened document suggestions must be greater that 0.

cmartel-coveo commented on packages/quantic/force-app/main/default/labels/CustomLabels.labels-meta.xml:
```diff
@@ -1106,4 +1106,67 @@
+    </labels>
+    <labels>
+      <fullName>quantic_InvalidMaxNumberOfSuggestions</fullName>
+      <value>"{{0}}" is an invalid maximum number of suggestions. A positive integer was expected.</value>
```
@mmitiche Au risque de me répéter... :) 
"The maximum number of suggestions must be greater than 0."

cmartel-coveo commented on packages/quantic/force-app/main/default/labels/CustomLabels.labels-meta.xml:
```diff
@@ -1106,4 +1106,67 @@
+    </labels>
+    <labels>
+      <fullName>quantic_PropertyIsRequired</fullName>
+      <value>The "{{0}}" property is required, please set its value.</value>
```
@mmitiche 
"Enter a value for "{{0}}"."

cmartel-coveo commented on packages/quantic/force-app/main/default/labels/CustomLabels.labels-meta.xml:
```diff
@@ -1106,4 +1106,67 @@
+      <value>Fatal error: unable to initialize the {{0}} component.</value>
+      <language>en_US</language>
+      <protected>false</protected>
+      <shortDescription>Fatal error: unable to initialize the component.</shortDescription>
```
@mmitiche 
Capitalize the first letter after the ":"

"Fatal error: Unable to initialize the component."

cmartel-coveo commented on packages/quantic/force-app/main/translations/fr.translation-meta.xml:
```diff
@@ -600,4 +600,40 @@
+        <name>quantic_LookAtDeveloperConsole</name>
+    </customLabels>
+    <customLabels>
+        <label>"{{0}}" est un nombre maximum de suggestions de documents non valide. Un nombre entier supérieur à 0 était attendu.</label>
```
@mmitiche If you decide to go with my EN suggestion, here's the FR equivalent:

"Le nombre maximum de suggestions de documents doit être supérieur à 0."

cmartel-coveo commented on packages/quantic/force-app/main/translations/fr.translation-meta.xml:
```diff
@@ -600,4 +600,40 @@
+        <name>quantic_SalesforceFieldNotFound</name>
+    </customLabels>
+    <customLabels>
+        <label>"{{0}}" est un nombre maximum de suggestions non valide. Un nombre entier positif était attendu.</label>
```
@mmitiche Again, if you take my EN suggestion:

"Le nombre maximum de suggestions doit être supérieur à 0."

cmartel-coveo commented on packages/quantic/force-app/main/translations/fr.translation-meta.xml:
```diff
@@ -600,4 +600,40 @@
+        <name>quantic_InvalidMaxNumberOfSuggestions</name>
+    </customLabels>
+    <customLabels>
+        <label>La propriété "{{0}}" est obligatoire, veuillez définir sa valeur.</label>
```
@mmitiche If EN suggestion is taken, FR equivalent is: 
"Il faut saisir une valeur pour la propriété "{{0}}"."

cmartel-coveo commented on packages/quantic/force-app/main/translations/fr.translation-meta.xml:
```diff
@@ -600,4 +600,40 @@
+        <name>quantic_ComponentInitializationError</name>
+    </customLabels>
+    <customLabels>
+        <label>Erreur fatale : impossible d'initialiser le composant {{0}}.</label>
```
@mmitiche 
J'irais aussi avec une majuscule après le deux-points:
"Erreur fatale : Impossible d'initialiser le composant {{0}}."

cmartel-coveo reviewed (commented):
@mmitiche I left comments where I saw potential improvements :) 
I left them as comments, as there are no major issues.

mmitiche commented on packages/quantic/force-app/main/default/labels/CustomLabels.labels-meta.xml:
```diff
@@ -1106,4 +1106,67 @@
+    </labels>
+    <labels>
+      <fullName>quantic_InvalidMaxNumberOfDocumentSuggestions</fullName>
+      <value>The maximum number of document suggestions must an integer be greater than 0.</value>
```
```suggestion
      <value>The maximum number of document suggestions must be an integer be greater than 0.</value>
```

mmitiche commented on packages/quantic/cypress/e2e/default-2/document-suggestion/document-suggestion.cypress.ts:
```diff
@@ -23,17 +23,11 @@ interface DocumentSuggestionOptions {
-  )}" is an invalid maximum number of document suggestions. A integer greater than 0 was expected.`;
-};
+const invalidMaxSuggestionsError =
+  'The maximum number of document suggestions must an integer be greater than 0.';
```
```suggestion
  'The maximum number of document suggestions must be an integer be greater than 0.';
```

cmartel-coveo commented on packages/quantic/force-app/main/default/labels/CustomLabels.labels-meta.xml:
```diff
@@ -1106,4 +1106,67 @@
+    </labels>
+    <labels>
+      <fullName>quantic_InvalidMaxNumberOfDocumentSuggestions</fullName>
+      <value>The maximum number of document suggestions must be an integer be greater than 0.</value>
```
@mmitiche Do you mean...?

"<value>The maximum number of document suggestions must be an **integer greater** than 0.</value>"

mmitiche replied:
yess exactly thank you for catching that 🙌

mmitiche commented on packages/quantic/force-app/main/default/labels/CustomLabels.labels-meta.xml:
```diff
@@ -1106,4 +1106,67 @@
+    </labels>
+    <labels>
+      <fullName>quantic_InvalidMaxNumberOfDocumentSuggestions</fullName>
+      <value>The maximum number of document suggestions must be an integer be greater than 0.</value>
```
```suggestion
      <value>The maximum number of document suggestions must be an integer greater than 0.</value>
```

mmitiche commented on packages/quantic/cypress/e2e/default-2/document-suggestion/document-suggestion.cypress.ts:
```diff
@@ -23,17 +23,11 @@ interface DocumentSuggestionOptions {
-  )}" is an invalid maximum number of document suggestions. A integer greater than 0 was expected.`;
-};
+const invalidMaxSuggestionsError =
+  'The maximum number of document suggestions must be an integer be greater than 0.';
```
```suggestion
  'The maximum number of document suggestions must be an integer greater than 0.';
```
